File: Classes/View Controllers/Tabs/Home/SearchSongsViewController.m
//
//  SearchSongsViewController.m
//  iSub
//
//  Created by bbaron on 10/28/10.
//  Copyright 2010 Ben Baron. All rights reserved.
//

#import "SearchSongsViewController.h"
#import "ServerListViewController.h"
#import "AlbumViewController.h"
#import "UIViewController+PushViewControllerCustom.h"
#import "SearchXMLParser.h"
#import "NSMutableURLRequest+SUS.h"
#import "ViewObjectsSingleton.h"
#import "Defines.h"
#import "SavedSettings.h"
#import "PlaylistSingleton.h"
#import "MusicSingleton.h"
#import "DatabaseSingleton.h"
#import "JukeboxSingleton.h"
#import "ISMSArtist.h"
#import "ISMSAlbum.h"
#import "ISMSSong+DAO.h"
#import "EX2Kit.h"
#import "Swift.h"
#import "SUSLoader.h"

@implementation SearchSongsViewController

#pragma mark View lifecycle

- (instancetype)initWithNibName:(NSString *)n bundle:(NSBundle *)b {
    if (self = [super initWithNibName:n bundle:b]) {
		_offset = 0;
		_isMoreResults = YES;
		_isLoading = NO;
    }
    return self;
}

- (void)viewDidLoad  {
    [super viewDidLoad];
	
	if (musicS.showPlayerIcon) {
		self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithImage:[UIImage systemImageNamed:@"music.quarternote.3"] style:UIBarButtonItemStylePlain target:self action:@selector(nowPlayingAction:)];
	} else {
		self.navigationItem.rightBarButtonItem = nil;
	}
    
    self.tableView.rowHeight = Defines.rowHeight;
    [self.tableView registerClass:UniversalTableViewCell.class forCellReuseIdentifier:UniversalTableViewCell.reuseId];
}
		
- (void)viewWillDisappear:(BOOL)animated {
	[super viewWillDisappear:animated];
	
	[self.connection cancel];
	self.connection = nil;
}

- (void) settingsAction:(id)sender  {
	ServerListViewController *serverListViewController = [[ServerListViewController alloc] initWithNibName:@"ServerListViewController" bundle:nil];
	serverListViewController.hidesBottomBarWhenPushed = YES;
	[self.navigationController pushViewController:serverListViewController animated:YES];
}

- (IBAction)nowPlayingAction:(id)sender {
    PlayerViewController *playerViewController = [[PlayerViewController alloc] init];
    playerViewController.hidesBottomBarWhenPushed = YES;
	[self.navigationController pushViewController:playerViewController animated:YES];
}

#pragma mark Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
	if (self.searchType == ISMSSearchSongsSearchType_Artists) {
		return self.listOfArtists.count + 1;
	} else if (self.searchType == ISMSSearchSongsSearchType_Albums) {
		return self.listOfAlbums.count + 1;
	} else {
		return self.listOfSongs.count + 1;
	}
}

- (void)loadMoreResults {
	if (self.isLoading) return;
	
	self.isLoading = YES;
	
	self.offset += 20;
    NSDictionary *parameters = nil;
    NSString *action = nil;
	NSString *offsetString = [NSString stringWithFormat:@"%lu", (unsigned long)self.offset];
	if (settingsS.isNewSearchAPI) {
        action = @"search2";
		NSString *queryString = [NSString stringWithFormat:@"%@*", self.query];
		switch (self.searchType) {
			case ISMSSearchSongsSearchType_Artists:
				parameters = [NSDictionary dictionaryWithObjectsAndKeys:@"20", @"artistCount", @"0", @"albumCount", @"0", @"songCount", 
							  n2N(queryString), @"query", n2N(offsetString), @"artistOffset", nil];
				break;
			case ISMSSearchSongsSearchType_Albums:
				parameters = [NSDictionary dictionaryWithObjectsAndKeys:@"0", @"artistCount", @"20", @"albumCount", @"0", @"songCount", 
							  n2N(queryString), @"query", n2N(offsetString), @"albumOffset", nil];
				break;
			case ISMSSearchSongsSearchType_Songs:
				parameters = [NSDictionary dictionaryWithObjectsAndKeys:@"0", @"artistCount", @"0", @"albumCount", @"20", @"songCount", 
							  n2N(queryString), @"query", n2N(offsetString), @"songOffset", nil];
				break;
			default:
				break;
		}
	} else {
        action = @"search";
        parameters = [NSDictionary dictionaryWithObjectsAndKeys:@"20", @"count", n2N(self.query), @"any", n2N(offsetString), @"offset", nil];
	}
    
    NSURLRequest *request = [NSMutableURLRequest requestWithSUSAction:action parameters:parameters];
    NSURLSessionDataTask *dataTask = [SUSLoader.sharedSession dataTaskWithRequest:request completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
        if (error) {
            [EX2Dispatch runInMainThreadAsync:^{
                if (settingsS.isPopupsEnabled) {
                    NSString *message = [NSString stringWithFormat:@"There was an error performing the search.\n\nError:%@", error.localizedDescription];
                    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Error" message:message preferredStyle:UIAlertControllerStyleAlert];
                    [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleCancel handler:nil]];
                    [self presentViewController:alert animated:YES completion:nil];
                }
                self.isLoading = NO;
            }];
        } else {
            SearchXMLParser *parserDelegate = [[SearchXMLParser alloc] init];
            NSXMLParser *xmlParser = [[NSXMLParser alloc] initWithData:self.receivedData];
            [xmlParser setDelegate:parserDelegate];
            [xmlParser parse];
            if (self.searchType == ISMSSearchSongsSearchType_Artists) {
                if (parserDelegate.listOfArtists.count == 0) {
                    self.isMoreResults = NO;
                } else {
                    [self.listOfArtists addObjectsFromArray:parserDelegate.listOfArtists];
                }
            } else if (self.searchType == ISMSSearchSongsSearchType_Albums) {
                if (parserDelegate.listOfAlbums.count == 0) {
                    self.isMoreResults = NO;
                } else {
                    [self.listOfAlbums addObjectsFromArray:parserDelegate.listOfAlbums];
                }
            } else if (self.searchType == ISMSSearchSongsSearchType_Songs) {
                if (parserDelegate.listOfSongs.count == 0) {
                    self.isMoreResults = NO;
                } else {
                    [self.listOfSongs addObjectsFromArray:parserDelegate.listOfSongs];
                }
            }
            
            // Reload the table
            [EX2Dispatch runInMainThreadAsync:^{
                [self.tableView reloadData];
                self.isLoading = NO;
            }];
        }
    }];
    [dataTask resume];
}

- (UITableViewCell *)createLoadingCell:(NSUInteger)row {
	// This is the last cell and there could be more results, load the next 20 results;
	UITableViewCell *cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"NoReuse"];
    cell.backgroundColor = [UIColor colorNamed:@"isubBackgroundColor"];
	if (self.isMoreResults) {
		cell.textLabel.text = @"Loading more results...";
        UIActivityIndicatorView *indicator = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleMedium];
		CGFloat y = [self tableView:self.tableView heightForRowAtIndexPath:[NSIndexPath indexPathForRow:row inSection:0]] / 2.;
		indicator.center = CGPointMake(300, y);
		indicator.autoresizingMask = UIViewAutoresizingFlexibleLeftMargin;
		[cell addSubview:indicator];
		[indicator startAnimating];
		
		[self loadMoreResults];
	} else {
		if (self.listOfArtists.count > 0 || self.listOfAlbums.count > 0 || self.listOfSongs.count > 0) {
			cell.textLabel.text = @"No more search results";
		} else {
			cell.textLabel.text = @"No results";
		}
	}
	return cell;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
	if (self.searchType == ISMSSearchSongsSearchType_Artists) {
		if (indexPath.row < self.listOfArtists.count) {
            // Artist
            UniversalTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:UniversalTableViewCell.reuseId];
            cell.hideNumberLabel = YES;
            cell.hideCoverArt = YES;
            cell.hideSecondaryLabel = YES;
            cell.hideDurationLabel = YES;
            [cell updateWithModel:[self.listOfArtists objectAtIndexSafe:indexPath.row]];
            return cell;
		} else if (indexPath.row == self.listOfArtists.count) {
			return [self createLoadingCell:indexPath.row];
		}
	} else if (self.searchType == ISMSSearchSongsSearchType_Albums) {
		if (indexPath.row < self.listOfAlbums.count) {
            // Album
            UniversalTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:UniversalTableViewCell.reuseId];
            cell.hideNumberLabel = YES;
            cell.hideCoverArt = NO;
            cell.hideDurationLabel = YES;
            [cell updateWithModel:[self.listOfAlbums objectAtIndexSafe:indexPath.row]];
            return cell;
		} else if (indexPath.row == [self.listOfAlbums count]) {
			return [self createLoadingCell:indexPath.row];
		}
	} else {
		if (indexPath.row < self.listOfSongs.count) {
            // Song
            UniversalTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:UniversalTableViewCell.reuseId];
            cell.hideNumberLabel = YES;
            cell.hideCoverArt = NO;
            cell.hideDurationLabel = NO;
            [cell updateWithModel:[self.listOfSongs objectAtIndexSafe:indexPath.row]];
            return cell;
		} else if (indexPath.row == self.listOfSongs.count) {
			return [self createLoadingCell:indexPath.row];
		}
	}
	
	// In case somehow no cell is created, return an empty cell
	static NSString *cellIdentifier = @"EmptyCell";
	return [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:cellIdentifier];
}

#pragma mark Table view delegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
	if (!indexPath) return;
	
	if (self.searchType == ISMSSearchSongsSearchType_Artists) {
		if (indexPath.row != self.listOfArtists.count) {
			ISMSArtist *anArtist = [self.listOfArtists objectAtIndexSafe:indexPath.row];
			AlbumViewController *albumView = [[AlbumViewController alloc] initWithArtist:anArtist orAlbum:nil];
			[self pushViewControllerCustom:albumView];
			return;
		}
	} else if (self.searchType == ISMSSearchSongsSearchType_Albums) {
		if (indexPath.row != self.listOfAlbums.count) {
			ISMSAlbum *anAlbum = [self.listOfAlbums objectAtIndexSafe:indexPath.row];
			AlbumViewController *albumView = [[AlbumViewController alloc] initWithArtist:nil orAlbum:anAlbum];
			[self pushViewControllerCustom:albumView];
			return;
		}
	} else {
		if (indexPath.row != self.listOfSongs.count) {
			// Clear the current playlist
			if (settingsS.isJukeboxEnabled) {
				[databaseS resetJukeboxPlaylist];
				[jukeboxS clearRemotePlaylist];
			} else {
				[databaseS resetCurrentPlaylistDb];
			}
			
			// Add the songs to the playlist 
			NSMutableArray *songIds = [[NSMutableArray alloc] init];
			for (ISMSSong *aSong in self.listOfSongs) {
				@autoreleasepool {
					[aSong addToCurrentPlaylistDbQueue];
					
					// In jukebox mode, collect the song ids to send to the server
                    if (settingsS.isJukeboxEnabled) {
                        [songIds addObject:aSong.songId];
                    }
				}
			}
			
			// If jukebox mode, send song ids to server
			if (settingsS.isJukeboxEnabled) {
				[jukeboxS stop];
				[jukeboxS clearPlaylist];
				[jukeboxS addSongs:songIds];
			}
			
			// Set player defaults
			playlistS.isShuffle = NO;
			
            [NSNotificationCenter postNotificationToMainThreadWithName:ISMSNotification_CurrentPlaylistSongsQueued];
            
			// Start the song
			ISMSSong *playedSong = [musicS playSongAtPosition:indexPath.row];
            if (!playedSong.isVideo) {
                [self showPlayer];
            }
			
			return;
		}
	}
	
	[self.tableView deselectRowAtIndexPath:indexPath animated:NO];
}

- (UISwipeActionsConfiguration *)tableView:(UITableView *)tableView trailingSwipeActionsConfigurationForRowAtIndexPath:(NSIndexPath *)indexPath {
    if (self.searchType == ISMSSearchSongsSearchType_Artists) {
        if (indexPath.row != self.listOfArtists.count) {
            return [SwipeAction downloadAndQueueConfigWithModel:[self.listOfArtists objectAtIndexSafe:indexPath.row]];
        }
    } else if (self.searchType == ISMSSearchSongsSearchType_Albums) {
        if (indexPath.row != self.listOfAlbums.count) {
            return [SwipeAction downloadAndQueueConfigWithModel:[self.listOfAlbums objectAtIndexSafe:indexPath.row]];
        }
    } else {
        if (indexPath.row != self.listOfSongs.count) {
            return [SwipeAction downloadAndQueueConfigWithModel:[self.listOfSongs objectAtIndexSafe:indexPath.row]];
        }
    }
    return nil;
}

@end

File: Classes/View Controllers/Tabs/Home/SearchSongsViewController.swift
//
//  SearchSongsViewController.swift
//  iSub
//
//  Created by Benjamin Baron on 1/19/21.
//  Copyright © 2021 Ben Baron. All rights reserved.
//

import UIKit
import Resolver
import CocoaLumberjackSwift

final class SearchSongsViewController: CustomUITableViewController {
    @Injected private var store: Store
    @Injected private var settings: SavedSettings
    @Injected private var jukebox: Jukebox
    @Injected private var playQueue: PlayQueue
    
    let serverId: Int
    let query: String
    let searchType: AsyncSearchLoader.SearchType
    let searchItemType: AsyncSearchLoader.SearchItemType
    
    private(set) var folderArtists: [FolderArtist]
    private(set) var folderAlbums: [FolderAlbum]
    private(set) var tagArtists: [TagArtist]
    private(set) var tagAlbums: [TagAlbum]
    private(set) var songs: [Song]
        
    private var offset = 0
    private var isMoreResults = true
    private var isLoading = false
//    private var searchLoader: SearchLoader
    private var loaderTask: Task<Void, Never>?
    
    init(serverId: Int, query: String, searchType: AsyncSearchLoader.SearchType, searchItemType: AsyncSearchLoader.SearchItemType, folderArtists: [FolderArtist] = [], folderAlbums: [FolderAlbum] = [], tagArtists: [TagArtist] = [], tagAlbums: [TagAlbum] = [], songs: [Song] = []) {
        self.serverId = serverId
        self.query = query
        self.searchType = searchType
        self.searchItemType = searchItemType
        self.folderArtists = folderArtists
        self.folderAlbums = folderAlbums
        self.tagArtists = tagArtists
        self.tagAlbums = tagAlbums
        self.songs = songs
//        self.searchLoader = SearchLoader(serverId: serverId, searchType: searchType, searchItemType: searchItemType, query: query)
        
        switch searchItemType {
        case .artists:
            let artists: [Any] = searchType == .folder ? folderArtists : tagArtists
            isMoreResults = artists.count >= AsyncSearchLoader.searchItemCount
        case .albums:
            let albums: [Any] = searchType == .folder ? folderAlbums : tagAlbums
            isMoreResults = albums.count >= AsyncSearchLoader.searchItemCount
        case .songs:
            isMoreResults = songs.count >= AsyncSearchLoader.searchItemCount
        default:
            isMoreResults = true
        }
        
        super.init(nibName: nil, bundle: nil)
    }
    
    required init?(coder: NSCoder) {
        fatalError("unimplemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        title = "Search Results"
        view.backgroundColor = Colors.background
        setupDefaultTableView(tableView)
    }
    
    deinit {
        loaderTask?.cancel()
//        searchLoader.cancelLoad()
//        searchLoader.callback = nil
    }
    
    private func loadMoreResults() {
        guard !isLoading else { return }
        isLoading = true
        
        offset += AsyncSearchLoader.searchItemCount
        loaderTask = Task {
            do {
                defer {
                    isLoading = false
                }
                
                let responseData = try await AsyncSearchLoader(serverId: serverId, searchType: searchType, searchItemType: searchItemType, query: query, offset: offset).load()
                switch self.searchItemType {
                case .artists:
                    if searchType == .folder {
                        folderArtists.append(contentsOf: responseData.folderArtists)
                        if responseData.folderArtists.count == 0 {
                            isMoreResults = false
                        }
                    } else {
                        tagArtists.append(contentsOf: responseData.tagArtists)
                        if responseData.tagArtists.count == 0 {
                            isMoreResults = false
                        }
                    }
                case .albums:
                    if searchType == .folder {
                        folderAlbums.append(contentsOf: responseData.folderAlbums)
                        if responseData.folderAlbums.count == 0 {
                            isMoreResults = false
                        }
                    } else {
                        tagAlbums.append(contentsOf: responseData.tagAlbums)
                        if responseData.tagAlbums.count == 0 {
                            isMoreResults = false
                        }
                    }
                case .songs:
                    songs.append(contentsOf: responseData.songs)
                    if responseData.songs.count == 0 {
                        isMoreResults = false
                    }
                default:
                    break
                }
                tableView.reloadData()
            } catch {
                if settings.isPopupsEnabled, !error.isCanceled {
                    self.isMoreResults = false
                    self.tableView.reloadData()
                    
                    // TODO: Verify if it's still necessary to use \(error.localizedDescription) or if I can just use \(error)
                    let message = "There was an error performing the search.\n\nError: \(error.localizedDescription)"
                    let alert = UIAlertController(title: "Error", message: message, preferredStyle: .alert)
                    alert.addOKAction()
                    self.present(alert, animated: true, completion: nil)
                }
            }
        }
    }
    
    override func tableCellModel(at indexPath: IndexPath) -> TableCellModel? {
        switch searchItemType {
        case .artists:
            let artists: [TableCellModel] = searchType == .folder ? folderArtists : tagArtists
            guard indexPath.row < artists.count else { return nil }
            return artists[indexPath.row]
        case .albums:
            let albums: [TableCellModel] = searchType == .folder ? folderAlbums : tagAlbums
            guard indexPath.row < albums.count else { return nil }
            return albums[indexPath.row]
        case .songs:
            guard indexPath.row < songs.count else { return nil }
            return songs[indexPath.row]
        default:
            return nil
        }
    }
}

extension SearchSongsViewController: UITableViewConfiguration {
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        switch searchItemType {
        case .artists: return (searchType == .folder ? folderArtists.count : tagArtists.count) + 1
        case .albums:  return (searchType == .folder ? folderAlbums.count : tagAlbums.count) + 1
        case .songs:   return songs.count + 1
        default:       return 0
        }
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        switch searchItemType {
        case .artists:
            let artists: [TableCellModel] = searchType == .folder ? folderArtists : tagArtists
            if indexPath.row < artists.count {
                let cell = tableView.dequeueUniversalCell()
                cell.show(downloaded: false, number: false, art: searchType == .tag, secondary: false, duration: false)
                cell.update(model: artists[indexPath.row])
                return cell
            }
        case .albums:
            let albums: [TableCellModel] = searchType == .folder ? folderAlbums : tagAlbums
            if indexPath.row < albums.count {
                let cell = tableView.dequeueUniversalCell()
                cell.show(downloaded: false, number: false, art: true, secondary: false, duration: false)
                cell.update(model: albums[indexPath.row])
                return cell
            }
        case .songs:
            if indexPath.row < songs.count {
                let cell = tableView.dequeueUniversalCell()
                cell.show(downloaded: true, number: false, art: true, secondary: true, duration: true)
                cell.update(model: songs[indexPath.row])
                return cell
            }
        default:
            break
        }
        
        // This is the last cell and there could be more results, load the next 20 songs;
        let cell = UITableViewCell(style: .default, reuseIdentifier: "SearchSongsLoadCell")
        cell.backgroundColor = .systemBackground
        if isMoreResults {
            cell.textLabel?.text = "Loading more results...";
            let indicator = UIActivityIndicatorView(style: .medium)
            indicator.center = CGPoint(x: 300, y: tableView.rowHeight / 2.0)
            indicator.autoresizingMask = .flexibleLeftMargin
            cell.addSubview(indicator)
            indicator.startAnimating()
            loadMoreResults()
        } else if folderArtists.count > 0 || folderAlbums.count > 0 || songs.count > 0 {
            cell.textLabel?.text = "No more search results"
        } else {
            cell.textLabel?.text = "No results"
        }
        
        handleOfflineMode(cell: cell, at: indexPath)
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        switch searchItemType {
        case .artists:
            if searchType == .folder {
                if indexPath.row < folderArtists.count {
                    let controller = FolderAlbumViewController(folderArtist: folderArtists[indexPath.row])
                    pushViewControllerCustom(controller)
                    return
                }
            } else {
                if indexPath.row < tagArtists.count {
                    let controller = TagArtistViewController(tagArtist: tagArtists[indexPath.row])
                    pushViewControllerCustom(controller)
                    return
                }
            }
        case .albums:
            if searchType == .folder {
                if indexPath.row < folderAlbums.count {
                    let controller = FolderAlbumViewController(folderAlbum: folderAlbums[indexPath.row])
                    pushViewControllerCustom(controller)
                    return
                }
            } else {
                if indexPath.row < tagAlbums.count {
                    let controller = TagAlbumViewController(tagAlbum: tagAlbums[indexPath.row])
                    pushViewControllerCustom(controller)
                    return
                }
            }
        case .songs:
            if indexPath.row < songs.count {
                if let song = store.playSong(position: indexPath.row, songs: songs), !song.isVideo {
                    NotificationCenter.postOnMainThread(name: Notifications.showPlayer)
                }
                return
            }
        default:
            break
        }
        
        // Loading cell
        tableView.deselectRow(at: indexPath, animated: false)
    }
    
    func tableView(_ tableView: UITableView, trailingSwipeActionsConfigurationForRowAt indexPath: IndexPath) -> UISwipeActionsConfiguration? {
        switch searchItemType {
        case .artists:
            let artists: [TableCellModel] = searchType == .folder ? folderArtists : tagArtists
            if indexPath.row < artists.count {
                return SwipeAction.downloadAndQueueConfig(model: artists[indexPath.row])
            }
        case .albums:
            let albums: [TableCellModel] = searchType == .folder ? folderAlbums : tagAlbums
            if indexPath.row < albums.count {
                return SwipeAction.downloadAndQueueConfig(model: albums[indexPath.row])
            }
        case .songs:
            if indexPath.row < songs.count {
                return SwipeAction.downloadAndQueueConfig(model: songs[indexPath.row])
            }
        default:
            break
        }
        return nil
    }
    
    func tableView(_ tableView: UITableView, contextMenuConfigurationForRowAt indexPath: IndexPath, point: CGPoint) -> UIContextMenuConfiguration? {
        switch searchItemType {
        case .artists:
            let artists: [TableCellModel] = searchType == .folder ? folderArtists : tagArtists
            if indexPath.row < artists.count {
                return contextMenuDownloadAndQueueConfig(model: artists[indexPath.row])
            }
        case .albums:
            let albums: [TableCellModel] = searchType == .folder ? folderAlbums : tagAlbums
            if indexPath.row < albums.count {
                return contextMenuDownloadAndQueueConfig(model: albums[indexPath.row])
            }
        case .songs:
            if indexPath.row < songs.count {
                return contextMenuDownloadAndQueueConfig(model: songs[indexPath.row])
            }
        default:
            break
        }
        return nil
    }
}
